vowpalwabbit/slates_label.cc

```diff
@@ -0,0 +1,209 @@
+  label.weight = 1.f;
+  label.labeled = false;
+  label.cost = 0.f;
+  label.slot_id = 0;
+  label.probabilities.clear();
+}
+
+bool test_label(void* v)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  return ld.labeled == false;
+}
+
+void delete_label(void* v)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  ld.probabilities.delete_v();
+}
+
+void copy_label(void* dst, void* src)
+{
+  auto& ldDst = static_cast<polylabel*>(dst)->slates;
+  auto& ldSrc = static_cast<polylabel*>(src)->slates;
+
+  ldDst.type = ldSrc.type;
+  ldDst.weight = ldSrc.weight;
+  ldDst.labeled = ldSrc.labeled;
+  ldDst.cost = ldSrc.cost;
+  ldDst.slot_id = ldSrc.slot_id;
+  copy_array(ldDst.probabilities, ldSrc.probabilities);
+}
+
+void parse_label(parser* p, shared_data*, void* v, v_array<VW::string_view>& words)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  ld.weight = 1;
+
+  if (words.size() < 2)
+    THROW("slates labels may not be empty");
+  if (!(words[0] == SLATES_LABEL))
+  {
+    THROW("slates labels require the first word to be slates");
+  }
+
+  const auto& type = words[1];
+  if (type == SHARED_TYPE)
+  {
+    // There is a cost defined.
+    if(words.size() == 3)
+    {
+      ld.cost = float_of_string(words[2]);
+      ld.labeled = true;
+    }
+    else if(words.size() != 2)
+    {
+      THROW("Slates shared labels must be of the form: slates shared [global_cost]");
+    }
+    ld.type = example_type::shared;
+  }
+  else if (type == ACTION_TYPE)
+  {
+    if(words.size() != 3)
+    {
+      THROW("Slates action labels must be of the form: slates action <slot_id>");
+    }
+
+    ld.slot_id = int_of_string(words[2]);
+    ld.type = example_type::action;
+  }
+  else if (type == SLOT_TYPE)
+  {
+    if (words.size() == 3)
+    {
+      ld.labeled = true;
+      tokenize(',', words[2], p->parse_name);
+
+      auto split_colons = v_init<VW::string_view>();
+      for(auto& token : p->parse_name) {
+        tokenize(':', token, split_colons);
+        if(split_colons.size() != 2)
+        {
+          THROW("Malformed action score token");
+        }
+
+        // Element 0 is the action, element 1 is the probability
+        ld.probabilities.push_back({static_cast<uint32_t>(int_of_string(split_colons[0])), float_of_string(split_colons[1])});
+      }
+
+      // If a full distribution has been given, check if it sums to 1, otherwise throw.
+      if (ld.probabilities.size() > 1)
+      {
+        float total_pred = std::accumulate(ld.probabilities.begin(), ld.probabilities.end(), 0.f,
+            [](float result_so_far, const ACTION_SCORE::action_score& action_pred) {
+              return result_so_far + action_pred.score;
+            });
+
+        // TODO do a proper comparison here.
+        if (total_pred > 1.1f || total_pred < 0.9f)
+        {
+          THROW("When providing all predicition probabilties they must add up to 1.f");
+        }
+      }
+    }
+    else if (words.size() != 2)
```

Comment:
Did you mean `== 2` or `< 3` here?

Reply:
The slot label can have either 2 or 3 tokens. Since at this point we know it doesn't have 3 because of the if statement, and outer checks mean it doesnt have 0 or 1. So the check should probably be > 3